src/Files.App/Helpers/GitHelpers.cs

```diff
@@ -252,22 +287,171 @@ public static async void PullOrigin(string? repositoryPath)
 			}
 			catch (Exception ex)
 			{
-				_logger.LogWarning(ex.Message);
+				if (ex.Message.Contains("status code: 401", StringComparison.OrdinalIgnoreCase) ||
+					ex.Message.Contains("authentication replays", StringComparison.OrdinalIgnoreCase))
+				{
+					await RequireGitAuthentication();
+				}
+				else
+				{
+					_logger.LogWarning(ex.Message);
+
+					var viewModel = new DynamicDialogViewModel()
+					{
+						TitleText = "GitError".GetLocalizedResource(),
+						SubtitleText = "PullTimeoutError".GetLocalizedResource(),
+						CloseButtonText = "Close".GetLocalizedResource(),
+						DynamicButtons = DynamicDialogButtons.Cancel
+					};
+					var dialog = new DynamicDialog(viewModel);
+					await dialog.TryShowAsync();
+				}
+			}
 
-				var viewModel = new DynamicDialogViewModel()
+			IsExecutingGitAction = false;
+		}
+
+		public static async Task PushToOrigin(string? repositoryPath, string? branchName)
+		{
+			if (string.IsNullOrWhiteSpace(repositoryPath) || string.IsNullOrWhiteSpace(branchName))
+				return;
+
+			using var repository = new Repository(repositoryPath);
+			var signature = repository.Config.BuildSignature(DateTimeOffset.Now);
+			if (signature is null)
+				return;
+
+			var token = CredentialsHelpers.GetPassword(GIT_RESOURCE_NAME, GIT_RESOURCE_USERNAME);
+			if (string.IsNullOrWhiteSpace(token))
+			{
+				await RequireGitAuthentication();
+				token = CredentialsHelpers.GetPassword(GIT_RESOURCE_NAME, GIT_RESOURCE_USERNAME);
+			}
+
+			var options = new PushOptions()
+			{
+				CredentialsProvider = (url, user, cred)
+					=> new UsernamePasswordCredentials
+					{
+						Username = signature.Name,
+						Password = token
+					}
+			};
+
+			IsExecutingGitAction = true;
+
+			try
+			{
+				var branch = repository.Branches[branchName];
+				if (!branch.IsTracking)
 				{
-					TitleText = "GitError".GetLocalizedResource(),
-					SubtitleText = "PullTimeoutError".GetLocalizedResource(),
-					CloseButtonText = "Close".GetLocalizedResource(),
-					DynamicButtons = DynamicDialogButtons.Cancel
-				};
-				var dialog = new DynamicDialog(viewModel);
-				await dialog.TryShowAsync();
+					var origin = repository.Network.Remotes["origin"];
+					repository.Branches.Update(
+						branch,
+						b => b.Remote = origin.Name,
+						b => b.UpstreamBranch = branch.CanonicalName);
+				}
+
+				repository.Network.Push(branch, options);
+			}
+			catch (Exception ex)
+			{
+				if (ex.Message.Contains("status code: 401", StringComparison.OrdinalIgnoreCase) ||
+					ex.Message.Contains("authentication replays", StringComparison.OrdinalIgnoreCase))
```

Comment:
Those message parts are used several times through the file. Fuse them in a test method.